sdks/java/io/pulsar/src/main/java/org/apache/beam/sdk/io/pulsar/PulsarIO.java

```diff
@@ -0,0 +1,146 @@
+            abstract Builder setEndMessageId(MessageId msgId);
+            abstract Builder setExtractOutputTimestampFn(SerializableFunction<Message<byte[]>, Instant> fn);
+            abstract Read build();
+        }
+
+
+        public Read withAdminUrl(String url) { return builder().setAdminUrl(url).build(); }
+
+        public Read withClientUrl(String url) {
+            return builder().setClientUrl(url).build();
+        }
+
+        public Read withTopic(String topic) {
+            return builder().setTopic(topic).build();
+        }
+
+        public Read withStartTimestamp(Long timestamp) {
+            return builder().setStartTimestamp(timestamp).build();
+        }
+
+        public Read withEndTimestamp(Long timestamp) {
+            return builder().setEndTimestamp(timestamp).build();
+        }
+
+        public Read withEndMessageId(MessageId msgId) {
+            return builder().setEndMessageId(msgId).build();
+        }
+
+        public Read withExtractOutputTimestampFn(SerializableFunction<Message<byte[]>, Instant> fn) {
+            return builder().setExtractOutputTimestampFn(fn).build();
+        }
+
+        public Read withPublishTime() {
+            return withExtractOutputTimestampFn(ExtractOutputTimestampFn.usePublishTime());
+        }
+
+        public Read withProcessingTime() {
+            return withExtractOutputTimestampFn(ExtractOutputTimestampFn.useProcessingTime());
+        }
+
+        @Override
+        public PCollection<PulsarMessage> expand(PBegin input) {
+            return input
+                    .apply(
+                            Create.of(
+                                    PulsarSourceDescriptor.of(getTopic(), getStartTimestamp(), getEndTimestamp(), getEndMessageId(), getClientUrl(), getAdminUrl())))
+                    .apply(
+                            ParDo.of(
+                                    new ReadFromPulsarDoFn(this)))
+                    .setCoder(PulsarMessageCoder.of());
+
+        }
+    }
+
+
+    public static Write write() {
+        return new AutoValue_PulsarIO_Write.Builder()
+                .build();
+    }
+
+    @AutoValue
+    @SuppressWarnings({"rawtypes"})
+    public abstract static class Write extends PTransform<PCollection<byte[]>, PDone> {
+
+        abstract @Nullable String getTopic();
+        abstract String getClientUrl();
+        abstract Builder builder();
+
+        @AutoValue.Builder
+        abstract static class Builder {
+            abstract Builder setTopic(String topic);
+            abstract Builder setClientUrl(String clientUrl);
+            abstract Write build();
+
+        }
+
+        public Write withTopic(String topic) {
+            return builder().setTopic(topic).build();
+        }
+
+        public Write withClientUrl(String clientUrl) {
+            return builder().setClientUrl(clientUrl).build();
+        }
+
+        @Override
+        public PDone expand(PCollection<byte[]> input) {
```

Comment:
Is byte[] the right type for this Write, or is there a higher level abstraction we could use?

Reply:
In order to send a message, by default Pulsar sends an array of bytes, since we don't have a way to know which type or schema end user will need, I thought it would be simple to leave as byte[] (See [Pulsar Producer client](https://pulsar.apache.org/docs/en/client-libraries-java/#producer))